File: models/src/agentii_models/options.py
"""Options models: Greeks, OptionQuote, OptionBar, OptionsChain.

Provider field mapping (Alpaca Snapshot → agentii-models):
    latestQuote.bp → bid, latestQuote.bs → bid_size,
    latestQuote.ap → ask, latestQuote.as → ask_size,
    latestTrade.p → last_trade_price, latestTrade.s → last_trade_size,
    greeks.delta → delta, greeks.gamma → gamma, greeks.theta → theta,
    greeks.vega → vega, greeks.rho → rho,
    impliedVolatility → implied_volatility, openInterest → open_interest

Provider field mapping (Polygon Snapshot → agentii-models):
    last_quote.bid → bid, last_quote.ask → ask,
    last_trade.price → last_trade_price, last_trade.size → last_trade_size,
    greeks.delta → delta, greeks.gamma → gamma, greeks.theta → theta,
    greeks.vega → vega, implied_volatility → implied_volatility,
    open_interest → open_interest
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, computed_field, field_validator

from ._base import BaseMarketData
from .enums import BarTimeframe, OptionType
from .types import Price, ProviderName, Symbol, Volume

_INTERDAY_TIMEFRAMES = frozenset({BarTimeframe.DAY_1, BarTimeframe.WEEK_1, BarTimeframe.MONTH_1})


class Greeks(BaseModel):
    """Position-level Greeks container for aggregation.
    Used by OptionPosition.position_greeks and Portfolio.net_greeks.
    NOT embedded on OptionQuote (those are flat fields)."""

    model_config = ConfigDict(frozen=True)

    delta: float = 0.0
    gamma: float = 0.0
    theta: float = 0.0
    vega: float = 0.0
    rho: float = 0.0
    implied_volatility: float = 0.0

    @computed_field  # type: ignore[prop-decorator]
    @property
    def is_complete(self) -> bool:
        return all(
            v != 0.0
            for v in (self.delta, self.gamma, self.theta, self.vega, self.rho, self.implied_volatility)
        )


class OptionQuote(BaseMarketData):
    """Real-time option contract quote with flat Greeks. Full OpenBB 35+ field set — all Optional."""

    # Identity
    underlying_symbol: Symbol
    contract_symbol: Symbol
    expiration: date
    dte: int | None = None
    strike: Price
    option_type: OptionType
    contract_size: int = 100

    # Pricing — bid/ask
    bid: Price | None = None
    bid_size: Volume | None = None
    bid_exchange: str | None = None
    bid_time: datetime | None = None
    ask: Price | None = None
    ask_size: Volume | None = None
    ask_exchange: str | None = None
    ask_time: datetime | None = None
    mark: Price | None = None

    # OHLC
    open: Price | None = None
    high: Price | None = None
    low: Price | None = None
    close: Price | None = None
    open_bid: Price | None = None
    open_ask: Price | None = None
    bid_high: Price | None = None
    ask_high: Price | None = None
    bid_low: Price | None = None
    ask_low: Price | None = None
    close_size: Volume | None = None
    close_time: datetime | None = None
    close_bid: Price | None = None
    close_bid_size: Volume | None = None
    close_bid_time: datetime | None = None
    close_ask: Price | None = None
    close_ask_size: Volume | None = None
    close_ask_time: datetime | None = None

    # Trade
    last_trade_price: Price | None = None
    last_trade_size: Volume | None = None
    last_trade_time: datetime | None = None
    tick: str | None = None

    # Change
    prev_close: Price | None = None
    change: float | None = None
    change_percent: float | None = None

    # Volume / OI
    volume: Volume | None = None
    open_interest: int | None = None

    # Greeks — FLAT on quote (not nested)
    implied_volatility: float | None = None
    delta: float | None = None
    gamma: float | None = None
    theta: float | None = None
    vega: float | None = None
    rho: float | None = None
    theoretical_price: Price | None = None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def mid_price(self) -> float | None:
        if self.bid is not None and self.ask is not None:
            return (self.bid + self.ask) / 2
        return None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def spread(self) -> float | None:
        if self.bid is not None and self.ask is not None:
            return self.ask - self.bid
        return None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def has_greeks(self) -> bool:
        return all(
            v is not None
            for v in (self.delta, self.gamma, self.theta, self.vega, self.rho, self.implied_volatility)
        )


class OptionBar(BaseMarketData):
    """OHLCV bar for a single option contract."""

    contract_symbol: Symbol
    underlying_symbol: Symbol
    date: date | datetime
    timeframe: BarTimeframe
    open: Price
    high: Price
    low: Price
    close: Price
    volume: Volume | None = None
    open_interest: int | None = None
    vwap: Price | None = None
    trade_count: int | None = None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def is_intraday(self) -> bool:
        return self.timeframe not in _INTERDAY_TIMEFRAMES


class OptionsChain(BaseModel):
    """Full chain snapshot for a single underlying.
    Stores contracts as list[OptionQuote] (per-contract objects)."""

    model_config = ConfigDict(frozen=True)

    underlying_symbol: Symbol
    snapshot_time: datetime
    provider: ProviderName
    contracts: list[OptionQuote]

    def filter(
        self,
        option_type: OptionType | None = None,
        expiration: date | None = None,
        min_strike: float | None = None,
        max_strike: float | None = None,
        min_volume: int | None = None,
        min_open_interest: int | None = None,
    ) -> list[OptionQuote]:
        result = self.contracts
        if option_type is not None:
            result = [c for c in result if c.option_type == option_type]
        if expiration is not None:
            result = [c for c in result if c.expiration == expiration]
        if min_strike is not None:
            result = [c for c in result if c.strike >= min_strike]
        if max_strike is not None:
            result = [c for c in result if c.strike <= max_strike]
        if min_volume is not None:
            result = [c for c in result if c.volume is not None and c.volume >= min_volume]
        if min_open_interest is not None:
            result = [c for c in result if c.open_interest is not None and c.open_interest >= min_open_interest]
        return result

    def by_expiry(self, expiration: date) -> list[OptionQuote]:
        return [c for c in self.contracts if c.expiration == expiration]

    def expirations(self) -> list[date]:
        return sorted({c.expiration for c in self.contracts})

    def strikes(self, expiration: date | None = None) -> list[float]:
        contracts = self.contracts if expiration is None else self.by_expiry(expiration)
        return sorted({c.strike for c in contracts})

    def atm_strike(self, underlying_price: float) -> float | None:
        all_strikes = self.strikes()
        if not all_strikes:
            return None
        return min(all_strikes, key=lambda s: abs(s - underlying_price))

    def near_term_expiry(self) -> date | None:
        today = date.today()
        future = [e for e in self.expirations() if e >= today]
        return future[0] if future else None

    def to_dataframe(self) -> Any:
        """Convert contracts to a pandas DataFrame. Requires pandas."""
        try:
            import pandas as pd
        except ImportError:
            raise ImportError("pandas is required for to_dataframe(). Install with: pip install pandas")
        return pd.DataFrame([c.model_dump() for c in self.contracts])

    @computed_field  # type: ignore[prop-decorator]
    @property
    def call_count(self) -> int:
        return sum(1 for c in self.contracts if c.option_type == OptionType.CALL)

    @computed_field  # type: ignore[prop-decorator]
    @property
    def put_count(self) -> int:
        return sum(1 for c in self.contracts if c.option_type == OptionType.PUT)

    @computed_field  # type: ignore[prop-decorator]
    @property
    def total_contracts(self) -> int:
        return len(self.contracts)

File: frontend/src/test/fixtures/options.ts
import type { OptionQuote, OptionsChain } from '@/types/options'

function generateOptionQuote(
  underlying: string,
  strike: number,
  expiration: string,
  type: 'call' | 'put',
  underlyingPrice: number,
): OptionQuote {
  const itm = type === 'call' ? underlyingPrice > strike : underlyingPrice < strike
  const moneyness = Math.abs(underlyingPrice - strike) / underlyingPrice
  const baseIV = 0.35 + moneyness * 0.5 + Math.random() * 0.05
  const dte = Math.max(1, Math.floor((new Date(expiration).getTime() - Date.now()) / 86400000))
  const timeValue = Math.sqrt(dte / 365) * underlyingPrice * baseIV * 0.4
  const intrinsic = itm ? Math.abs(underlyingPrice - strike) : 0
  const mid = intrinsic + timeValue
  const spread = Math.max(0.05, mid * 0.03)

  const delta = type === 'call'
    ? 0.5 + (underlyingPrice - strike) / (underlyingPrice * 0.3) * 0.3
    : -(0.5 - (underlyingPrice - strike) / (underlyingPrice * 0.3) * 0.3)

  const padSymbol = underlying.padEnd(6, ' ')
  const yr = expiration.slice(2, 4)
  const mo = expiration.slice(5, 7)
  const dy = expiration.slice(8, 10)
  const typeChar = type === 'call' ? 'C' : 'P'
  const strikeStr = String(Math.round(strike * 1000)).padStart(8, '0')
  const contractSymbol = `${padSymbol}${yr}${mo}${dy}${typeChar}${strikeStr}`

  return {
    symbol: underlying,
    provider: 'mock',
    timestamp_ns: null,
    underlying_symbol: underlying,
    contract_symbol: contractSymbol,
    expiration,
    dte,
    strike,
    option_type: type,
    contract_size: 100,
    bid: +Math.max(0.01, mid - spread / 2).toFixed(2),
    bid_size: Math.floor(Math.random() * 50) + 1,
    bid_exchange: null,
    bid_time: null,
    ask: +(mid + spread / 2).toFixed(2),
    ask_size: Math.floor(Math.random() * 50) + 1,
    ask_exchange: null,
    ask_time: null,
    mark: +mid.toFixed(2),
    open: null, high: null, low: null, close: null,
    open_bid: null, open_ask: null, bid_high: null, ask_high: null,
    bid_low: null, ask_low: null, close_size: null, close_time: null,
    close_bid: null, close_bid_size: null, close_bid_time: null,
    close_ask: null, close_ask_size: null, close_ask_time: null,
    last_trade_price: +mid.toFixed(2),
    last_trade_size: 10,
    last_trade_time: null,
    tick: null,
    prev_close: +(mid * 0.98).toFixed(2),
    change: +(mid * 0.02).toFixed(2),
    change_percent: 2.0,
    volume: Math.floor(Math.random() * 5000) + 100,
    open_interest: Math.floor(Math.random() * 20000) + 500,
    implied_volatility: +baseIV.toFixed(4),
    delta: +Math.max(-1, Math.min(1, delta)).toFixed(4),
    gamma: +(0.02 + Math.random() * 0.03).toFixed(4),
    theta: +(-0.01 - Math.random() * 0.05).toFixed(4),
    vega: +(0.05 + Math.random() * 0.1).toFixed(4),
    rho: +(0.01 + Math.random() * 0.02).toFixed(4),
    theoretical_price: +mid.toFixed(2),
    mid_price: +mid.toFixed(2),
    spread: +spread.toFixed(2),
    has_greeks: true,
  }
}

export function generateOptionsChain(symbol: string, underlyingPrice: number): OptionsChain {
  const expirations = ['2026-03-21', '2026-04-17', '2026-05-15', '2026-06-19', '2026-09-18']
  const contracts: OptionQuote[] = []

  for (const exp of expirations) {
    const atmStrike = Math.round(underlyingPrice / 5) * 5
    for (let offset = -10; offset <= 10; offset++) {
      const strike = atmStrike + offset * 5
      if (strike <= 0) continue
      contracts.push(generateOptionQuote(symbol, strike, exp, 'call', underlyingPrice))
      contracts.push(generateOptionQuote(symbol, strike, exp, 'put', underlyingPrice))
    }
  }

  return {
    underlying_symbol: symbol,
    snapshot_time: new Date().toISOString(),
    provider: 'mock',
    contracts,
    call_count: contracts.filter((c) => c.option_type === 'call').length,
    put_count: contracts.filter((c) => c.option_type === 'put').length,
    total_contracts: contracts.length,
  }
}

export const mrnaOptionsChain = generateOptionsChain('MRNA', 42)
